src/networking.c

```diff
@@ -4316,44 +4317,47 @@ void killIOThreads(void) {
     }
 }
 
-void startThreadedIO(void) {
-    serverAssert(server.io_threads_active == 0);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_unlock(&io_threads_mutex[j]);
-    server.io_threads_active = 1;
-}
-
-void stopThreadedIO(void) {
-    /* We may have still clients with pending reads when this function
-     * is called: handle them before stopping the threads. */
-    handleClientsWithPendingReadsUsingThreads();
-    serverAssert(server.io_threads_active == 1);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_lock(&io_threads_mutex[j]);
-    server.io_threads_active = 0;
-}
-
-/* This function checks if there are not enough pending clients to justify
- * taking the I/O threads active: in that case I/O threads are stopped if
- * currently active. We track the pending writes as a measure of clients
- * we need to handle in parallel, however the I/O threading is disabled
- * globally for reads as well if we have too little pending clients.
+/* This function dynamically adjusts the I/O thread number according to
+ * real workloads in runtime, currently we track the latest x rounds of
+ * pending writes as a measure of clients we need to handle in parallel,
+ * however the I/O threading is disabled globally for reads as well if
+ * we have too little pending clients.
  *
  * The function returns 0 if the I/O threading should be used because there
  * are enough active threads, otherwise 1 is returned and the I/O threads
  * could be possibly stopped (if already active) as a side effect. */
+#define IO_THREAD_BATCH_NUM 2
+#define SMOOTH_FACTOR 0.8
 int stopThreadedIOIfNeeded(void) {
```

Comment:
Sorry for going back and forth on this one but on a second thought, I think your previous function name makes more sense. 

```suggestion
int adjustIOThreadCount(void) {
```